packages/appium/lib/extension/manifest.js

```diff
@@ -0,0 +1,564 @@
+      if (!this._data.drivers[pkgJson.appium.driverName]) {
+        this._data.drivers[pkgJson.appium.driverName] = {
+          ..._.omit(pkgJson.appium, 'driverName'),
+          ...internal,
+        };
+      }
+    } else if (isPlugin(pkgJson)) {
+      if (!this._data.plugins[pkgJson.appium.pluginName]) {
+        this._data.plugins[pkgJson.appium.pluginName] = {
+          ..._.omit(pkgJson.appium, 'pluginName'),
+          ...internal,
+        };
+      }
+    } else {
+      throw new TypeError(
+        `The extension in ${path.dirname(
+          pkgPath,
+        )} is neither a valid driver nor a valid plugin.`,
+      );
+    }
+  }
+
+  /**
+   * Adds an extension to the manifest as was installed by the `appium` CLI.  It determines the
+   * `extData`, `extType`, `extName` itself.
+   *
+   * See {@link Manifest.addExtensionFromPackage} for adding an extension from an on-disk package.
+   * @template {ExtensionType} ExtType
+   * @param {ExtType} extType
+   * @param {string} extName
+   * @param {ExtData<ExtType>} extData
+   * @returns {void}
+   */
+  addExtension (extType, extName, extData) {
+    this._data[`${extType}s`][extName] = extData;
+  }
+
+  /**
+   * Returns the APPIUM_HOME path
+   */
+  get appiumHome () {
+    return this._appiumHome;
+  }
+
+  /**
+   * Returns the path to the manifest file
+   */
+  get manifestPath () {
+    return this._manifestPath;
+  }
+
+  /**
+   * Returns extension data for a particular type.
+   *
+   * @template {ExtensionType} ExtType
+   * @param {ExtType} extType
+   * @returns {ExtRecord<ExtType>}
+   */
+  getExtensionData (extType) {
+    return this._data[/** @type {string} */ (`${extType}s`)];
+  }
+
+  /**
+   * Reads manifest from disk and _overwrites_ the internal data.
+   *
+   * If the manifest does not exist on disk, an {@link INITIAL_MANIFEST_DATA "empty"} manifest file will be created.
+   *
+   * If `APPIUM_HOME` contains a `package.json` with an `appium` dependency, then a hash of the `package.json` will be taken. If this hash differs from the last hash, the contents of `APPIUM_HOME/node_modules` will be scanned for extensions that may have been installed outside of the `appium` CLI.  Any found extensions will be added to the manifest file, and if so, the manifest file will be written to disk.
+   *
+   * Only one read operation should happen at a time.  This is controlled via the {@link Manifest._reading} property.
+   * @returns {Promise<ManifestData>} The data
+   */
+  async read () {
+    if (this._reading) {
+      await this._reading;
+      return this._data;
+    }
+
+    this._reading = (async () => {
+      /** @type {ManifestData} */
+      let data;
+      let isNewFile = false;
+      await this._setManifestPath();
+      try {
+        log.debug(`Reading ${this._manifestPath}...`);
+        const yaml = await fs.readFile(this._manifestPath, 'utf8');
+        data = YAML.parse(yaml);
+      } catch (err) {
+        if (err.code === 'ENOENT') {
```

Comment:
I would rather prefer to explicitly check if the file exists first. Exception-driven flow is not very obvious

Reply:
I learned the exact opposite.  checking if a file exists first is more I/O overhead